SliderButton/SliderButtonBehaviorDecorator.js

```diff
@@ -10,43 +12,71 @@ import {findDOMNode} from 'react-dom';
  * @memberof agate/SliderButton
  * @private
  */
-const SliderButtonBehaviorDecorator = (Wrapped) => {
-	// eslint-disable-next-line no-shadow
-	function SliderButtonBehaviorDecorator ({onChange, ...rest}) {
-		const {children} = rest;
-		const [valueText, setValueText] = React.useState(children ? children[0] : null);
-		const ref = React.useRef();
-
-		function handleChange ({value}) {
-			onChange();
-			setValueText(children[value]);
+const SliderButtonBehaviorDecorator = hoc((config, Wrapped) => {
+	return class extends React.Component {
+		static displayName = 'SliderButtonBehaviorDecorator';
+
+		static propTypes = /** @lends agate/ArcSlider.ArcSliderBehaviorDecorator.prototype */ {
+			/**
+			 * Items displayed with SliderButton.
+			 *
+			 * @type {Function}
+			 * @private
+			 */
+			children: PropTypes.node,
+
+			/**
+			 * Called when value is changed.
+			 *
+			 * @type {Function}
+			 * @private
+			 */
+			onChange: PropTypes.func
+		};
+
+		constructor (props) {
+			super(props);
+
+			this.ref = React.createRef();
+
+			this.state = {
+				valueText: props.children ? props.children[0] : null
+			};
 		}
 
-		function handleDragStart () {
+		handleChange = ({value}) => {
+			this.setState(
+				() => ({valueText: this.props.children[value]}),
```

Comment:
Is there any reason to update the `valueText` state before calling `onChange` prop? I don't know why we didn't before exactly.

Reply:
No particular reason, I saw it implemented the same way in other component. Do you think it should be changed?

Reply:
I'll discuss it with our members and let you know :)

Reply:
@MikyungKim , @0x64 , @jeonghee27 
Please share your opinions for the following points.

We usually call a props's even handler and setState() together in an event handler.
I want to discuss what is our preference for the following things.
- function vs object as the first argument in the SetState
My opinion:
If we need to refer the previous state or props or to call other functions after updating states, we need to use the function. In other cases, we could use the object simply.

- Which one between a prop's event handler and setState() should be called first?
My opinion:
If we call the `setState()` to update any status or any values which can be referred by apps, we need to call the prop's event handler after updating the state. Otherwise, I think we don't care including the case updating just a view.

### Original code
```
		function handleChange ({value}) {
			onChange();
			setValueText(children[value]);
		}
```
My opinion:
The `children` is cached as closer above and is not props or previous state. So it's ok to pass an object as the first argument of the `setValueText`.
I'm fine to call `onChange` first or after calling `setValueText`.

### The first patch
```
		handleChange = ({value}) => {
			this.setState(
				() => ({valueText: this.props.children[value]}),
				() => {
					forward('onChange', {
						type: 'onChange',
						value
					}, this.props);
				}
			);
		};
```
`(prevState, props) => ({valueText: props.children[value]}),` would be suitable than `() => ({valueText: this.props.children[value]}),` because of using proper props.
`forward` would be out of the `setState` because we called `setState` to update a view so that app developers could handle it early.

### The current patch
```
		handleChange = ({value}) => {
			this.setState(() => ({valueText: this.props.children[value]}));
			forward('onChange', {
				type: 'onChange',
				value
			}, this.props);
		};
```
If converting `this.setState(() => ({valueText: this.props.children[value]}));` to `this.setState((prevState, props) => ({valueText: props.children[value]}),);`, everything else is good to me.

Reply:
> function vs object as the first argument in the SetState
I agree with @ybsung. If it is not required to access previous values, passing an object looks much better.

> Which one between a prop's event handler and setState() should be called first? 
In this case, `forward` handles _ev_ object anyway, so it looks not so important here. I don't have any preference of orders between them.

Reply:
Hi @adrian-cocoara-lgp 

Please check our discussion above and the following comment
> If converting this.setState(() => ({valueText: this.props.children[value]})); to this.setState((prevState, props) => ({valueText: props.children[value]}),);, everything else is good to me.

So my suggestion is..
```
		handleChange = ({value}) => {
			this.setState((prevState, props) => ({valueText: props.children[value]}));
			forward('onChange', {
				type: 'onChange',
				value
			}, this.props);
		};
```

Reply:
Thanks for detailed answers. I will change this accordingly.